### PR TITLE
tgi driver for c128 VIC-II

### PR DESCRIPTION
**Pull request by xlar54**
My first attempt at a driver for cc65, but trying to give back to you guys.  Hope its ok.

### CODE REVIEW

**Comment by polluks** [comment]
VBASE $6000? You cannot use `GRAPHICS1:GRAPHICS0`to move the BASIC start anymore.
cc65 formatting rule: No tabs!

**Comment by xlar54** [comment]
Yeah, Im definitely new to VIC programming on the 128, so Im sure there are things in there than should be adjusted or accounted for.  But the code works as is at the moment, so please do use it and change as appropriately needed.

**Comment by greg-king5** [comment]
cc65 programs use C128 bank zero memory from $1C00 to $BFFF.

Originally, I thought that a driver would need to do things the way that the BASIC interpreter does it (move the program to a higher address, to make enough space for graphics).  But now, I realize that it can put the bitmap under the screen-editor ROM and the I/O chips ($C000 in bank 0).  It can put the color map under the Kernal ROM ($E000).

The driver must use the MMU to disable the ROMs -- location $01 works differently than it does on the C64.

**Comment by xlar54** [comment]
Honestly, I was rather surprised that i got it to work (even though we may want to move things).  Im also using the shadow registers, so that potentially the split screen options might be used as well.  You guys have done a great job making this extensible and I just wanted to put that out there.  The code is clean and easy to work with.

**Comment by oliverschmidt** [comment]
> You guys have done a great job making this extensible and I just wanted to put that out there. The code is clean and easy to work with.

Maybe I'm missing the point but as far as I can tell the code is AS-IS not ready to be merged. And that's not only because of the tab chars in the source ;-)

At least I personally am not willing to work with it to get it into a state of being ready to be merged. The usual cc65 work flow is that the person creating a pull request moves it forward until it can be merged...

**Comment by xlar54** [comment]
No, that comment was meant to be a compliment to *your* code, not mine.  Im aware of the tab issue, but what else is needed to be fixed (unless you are referring to greg's comments about moving it to another location).  Thanks!  Process is understood. Just been a little busy with another project.

**Comment by oliverschmidt** [comment]
Thanks for the clarification! I just wanted to make sure there's no mismatch of expectations.

**Comment by xlar54** [comment]
No worries!

**Comment by polluks** [comment]
@xlar54 A new section in c128.sgml.

**Comment by xlar54** [comment]
Ill have to retract my pull request then.  Scouring over the details of the MMU is just terribly confusing and over my head for now.  But I am eager to see how it turns out.  Thanks guys!

**Comment by oliverschmidt** [comment]
I see. Then just leave this pr as-is.

Anyhow, thanks for your willingness to contribute to cc65.

**Comment by polluks** [comment]
BTW this works:
`dload"darth-vader"
run:rem c128-hi.tgi`

http://sourceforge.net/projects/cc65/files/contrib/darth-vader.zip/download

**Comment by xlar54** [comment]
yeah I think the code is good, its just that it seems we want to move it to a different location, and that MMU is a scary animal. (for me anyway)

**Comment by greg-king5** [comment]
You should merge the two Pull Requests that Stefan and I submitted to your GitHub fork.  I believe that they will make this PR mergeable.

**Comment by xlar54** [comment]
ok, will work on that this week

**Comment by xlar54** [comment]
Works like a champ, thanks Greg. Ill be studying your changes to understand them better.

File: libsrc/c128/tgi/c128-hi.s [hunk 1]
```diff
@@ -0,0 +1,880 @@
...
+; Switch into graphics mode.
+
+; Select a video bank:
+; bank 0 = $03 ($0000-$3FFF) (default)
+; bank 1 = $02 ($4000-$7FFF)
+; bank 2 = $01 ($8000-$BFFF)
+; bank 3 = $00 ($C000-$FFFF) (TGI)
+
+        lda     CIA2_PRA
+        and     #<~$03          ; Bank 3
```
Comment (oliverschmidt) [inline libsrc/c128/tgi/c128-hi.s]: Is it necessary/desirable to explicitly select the low byte here?

File: libsrc/c128/tgi/c128-hi.s [hunk 2]
```diff
@@ -0,0 +1,880 @@
...
+        ldx     #((>(CHARROM + $0000)) >> 3)
+@L1:    stx     ptr4+1
+        asl     a
+        rol     ptr4+1
+        asl     a
+        rol     ptr4+1
+        asl     a
+        rol     ptr4+1
+        sta     ptr4
+
```
Comment (oliverschmidt) [inline libsrc/c128/tgi/c128-hi.s]: I don't think that these blank lines belong here.

**Comment by oliverschmidt** [comment]
> Is there anything more I need to do to merge this into main?

Wait for Greg to assess it / merge it, I'd say.

File: libsrc/c128/tgi/c128-hi.s (hunk 2 above)
Comment (polluks) [reply libsrc/c128/tgi/c128-hi.s]: This part is everywhere incomplete, we should insert a TODO.

File: libsrc/c128/tgi/c128-hi.s (hunk 1 above)
Comment (greg-king5) [reply libsrc/c128/tgi/c128-hi.s]: We get a range error if we don't use `<`.  Ca65 expressions are calculated in the width (32/64 bits) of the `signed long` data type of the compiler that builds ca65.  Currently, the values aren't demoted automatically to the width of 6502 operands -- we must do it with the proper operator or function.

Comment (oliverschmidt) [reply libsrc/c128/tgi/c128-hi.s]: Thanks for the explanation :-)

File: libsrc/c128/tgi/c128-hi.s (hunk 2 above)
Comment (oliverschmidt) [reply libsrc/c128/tgi/c128-hi.s]: I see.

**Comment by oliverschmidt** [comment]
@xlar54 & @greg-king5 Thanks :-)

**Pull request merged**